chain/consensus/mir/mir_agent.go

```diff
@@ -0,0 +1,163 @@
+package mir
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/mir"
+	mirCrypto "github.com/filecoin-project/mir/pkg/crypto"
+	"github.com/filecoin-project/mir/pkg/grpctransport"
+	"github.com/filecoin-project/mir/pkg/iss"
+	mirLogging "github.com/filecoin-project/mir/pkg/logging"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/reqstore"
+	"github.com/filecoin-project/mir/pkg/simplewal"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+// agentLog is a logger accessed by Mir.
+var agentLog = logging.Logger("mir-agent")
+
+// mirLogger implement Mir's Log interface.
+type mirLogger struct {
+	logger *logging.ZapEventLogger
+}
+
+func newMirLogger(logger *logging.ZapEventLogger) *mirLogger {
+	return &mirLogger{
+		logger: logger,
+	}
+}
+
+// Log logs a message with additional context.
+func (m *mirLogger) Log(level mirLogging.LogLevel, text string, args ...interface{}) {
+	switch level {
+	case mirLogging.LevelError:
+		m.logger.Errorw(text, args)
+	case mirLogging.LevelInfo:
+		m.logger.Infow(text, args)
+	case mirLogging.LevelWarn:
+		m.logger.Warnw(text, args)
+	case mirLogging.LevelDebug:
+		m.logger.Debugw(text, args)
+	}
+}
+
+// MirAgent manages and provides direct access to a Mir node abstraction participating in consensus.
+type MirAgent struct {
+	Node     *mir.Node
+	Wal      *simplewal.WAL
+	Net      *grpctransport.GrpcTransport
+	App      *Application
+	stopChan chan struct{}
+}
+
+func NewMirAgent(id string, n int) (*MirAgent, error) {
+	// TODO: Are client ID and node ID the same in this case?
```

Comment:
@matejpavlovic? This is more a conceptual question, right?

Reply:
Copy-pasting from my answer to the previous comment:
> If I understood correctly from previous discussions, each message has a "nonce", which is a monotonically increasing number (I'd prefer calling it a "message number", "request number", or "sequence number", but as long as we all know what we're talking about, I don't care that much). This nonce uniquely identifies a message among all the messages coming, in a way, from the same "source". This source is something that produces a sequence of messages and assigns them the nones. In other words, the source defines a set of messages among which nonces are unique. The identifier of this source is what we need to use as "client ID".

As for node ID, it can be anything that uniquely identifies a node that participates in the consensus protocol. The Eudico node's ID seems like a natural choice to me here.

Reply:
We discussed with @adlrocha : subnetName + wallet address can be used as IDs.

Reply:
Sounds reasonable.